packages/messaging/src/interfaces/firebase-messaging.ts

```diff
@@ -14,31 +14,32 @@
  * limitations under the License.
  */
 
-import { FirebaseApp, FirebaseNamespace } from '@firebase/app-types';
 import {
-  Observer,
-  Unsubscribe,
-  NextFn,
+  CompleteFn,
   ErrorFn,
-  CompleteFn
+  NextFn,
+  Observer,
+  Unsubscribe
 } from '@firebase/util';
 
-export class FirebaseMessaging {
-  private constructor();
+export interface FirebaseMessaging {
   deleteToken(token: string): Promise<boolean>;
   getToken(): Promise<string | null>;
   onMessage(
+    // tslint:disable-next-line no-any The message payload can be anything.
```

Comment:
Isn't object the better type for "I have no idea what this is, but I don't do anything specific with it"? And that's what the implementation seems to use anyway?

https://github.com/firebase/firebase-js-sdk/blob/9fd02293703c20a9c624f28ec6b54325f67e9ffa/packages/messaging/src/controllers/window-controller.ts#L162

Reply:
The implementation uses `object` because inside the SDK we don't (and shouldn't) access any properties of the payload. Like you said, we don't do anything specific with it.

The exported types are different, because the user should know and be able to access their payload's properties.

So this would work with `any` and wouldn't work with `object` (when `--strict` flag is on, because of [strictFunctionTypes](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-6.html)):

```
messaging.onMessage((payload: UserDefinedPayloadObjectType) => {
  const messageData = payload.someField;
  // Do stuff with payload.
});
```

I'm planning on eventually making this a [MessagePayload](https://github.com/firebase/firebase-js-sdk/blob/4ff990a9f0a0d804a63bcd49f22e2cd86a216721/packages/messaging/src/interfaces/message-payload.ts#L22) and preventing the `--strict` errors, but I'm about to propose some changes to that type, so I think it's better to keep this as `any` for now and change it later.

Reply:
Right, the type safety escape hatch is used because this isn't type safe, gotcha :/

Seems to me like the whole FirebaseMessaging interface should take a type parameter of the expected type, and then callers are expected to provide callbacks that handle precisely that type. In the meantime, carry on I guess :/